src/components/Header/index.js

```diff
@@ -17,9 +19,37 @@ import css from './Header.scss'
 const cx = className.bind(css)
 
 class Header extends Component {
+  componentDidMount() {
+    if (this.props.isTournament && this.props.currentAccount) {
+      this.props.requestMainnetAddress()
+    }
+  }
+
+  componentDidUpdate(prevProps) {
+    // If user unlocks metamask, changes his account, we need to check if the account was registered
+    const shouldRequestMainnetAddress =
+      this.props.isTournament && this.props.currentAccount !== prevProps.currentAccount
+    if (shouldRequestMainnetAddress) {
+      this.props.requestMainnetAddress()
+    }
+  }
+
   @autobind
-  handleConnectWalletClick() {
-    this.props.openConnectWalletModal()
+  async handleConnectWalletClick() {
+    const { isConnectedToCorrectNetwork, lockedMetamask } = this.props
+    if (!hasMetamask() && !shouldUseUport) {
+      this.props.openInstallMetamaskModal()
+    } else if (shouldUseMetamask()) {
+      if (lockedMetamask) {
+        this.props.openUnlockMetamaskModal()
+      } else if (!isConnectedToCorrectNetwork) {
+        this.props.openSwitchNetworkModal()
+      } else {
+        this.props.openRegisterWalletModal()
```

Comment:
I think what would be nice here for all the modal functions, would be a function called `openModal` where you simply pass the name of the modal (and additional parameters to display it) instead of having very verbose action functions for each modal. What do you think?